src/mesido/workflows/multicommodity_simulator_workflow.py

```diff
@@ -0,0 +1,624 @@
+
+
+class MaximizeDemandGoalMerit(Goal):
+    """
+    Apply constraints to enforce esdl specified milp producer merit order usage
+    """
+
+    def __init__(self, demand_variable, prod_priority, func_range_bound, nominal, order=2):
+
+        self.demand_variable = demand_variable
+        self.function_nominal = nominal
+        self.priority = prod_priority
+        self.order = 1
+
+    def function(self, optimization_problem, ensemble_member):
+        return -optimization_problem.state(f"{self.demand_variable}")
+
+
+class MinimizeStorageGoalMerit(Goal):
+    """
+    Apply constraints to enforce esdl specified milp producer merit order usage
+    """
+
+    def __init__(self, source_variable, prod_priority, func_range_bound, nominal, order=2):
+        self.target_min = 0.0  # func_range_bound[0]
+        self.target_max = func_range_bound[1] * 0.999
+        self.function_range = func_range_bound
+        self.source_variable = source_variable
+        self.function_nominal = nominal
+        self.priority = prod_priority
+        self.order = order
+
+    def function(self, optimization_problem, ensemble_member):
+        return optimization_problem.state(f"{self.source_variable}")
+
+
+class MaximizeStorageGoalMerit(Goal):
+    """
+    Apply constraints to enforce esdl specified milp producer merit order usage
+    """
+
+    def __init__(self, demand_variable, prod_priority, func_range_bound, nominal, order=2):
+        self.target_min = func_range_bound[1]
+        self.function_range = func_range_bound
+        self.demand_variable = demand_variable
+        self.function_nominal = nominal
+        self.priority = prod_priority + 1
+        self.order = order
+
+    def function(self, optimization_problem, ensemble_member):
+        return optimization_problem.state(f"{self.demand_variable}")
+
+
+# -------------------------------------------------------------------------------------------------
+class _GoalsAndOptions:
+    def path_goals(self):
+        goals = super().path_goals().copy()
+
+        # Demand and producer matching
+        map_demand = {
+            "electricity_demand": {
+                "target": "target_electricity_demand",
+                "state": "Electricity_demand",
+            },
+            "gas_demand": {"target": "target_gas_demand", "state": "Gas_demand_mass_flow"},
+        }
+        map_prod = {
+            "electricity_source": {
+                "target": "maximum_electricity_source",
+                "state": "Electricity_source",
+            },
+            "gas_source": {"target": "maximum_gas_source", "state": "Gas_source_mass_flow"},
+        }
+        for type, type_values in [*map_demand.items(), *map_prod.items()]:
+            for asset in self.energy_system_components.get(type, []):
+                timeseries_name = f"{asset}.{type_values['target']}"
+                if timeseries_name in self.io.get_timeseries_names():
```

Comment:
The goal should only be added in case that a timeseries exists for the asset. The name of the timeseries is standardised, for every asset type, there might exist other timeseries but those are not relevant for these goals.